### DISCUSSION TITLE
GitLab npm removing packages from package-lock.json

### DISCUSSION THREAD

**Discussion by juancarlosjr97** [renovatebot/renovate#14985] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v32.11.3

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

v32.11.3

### Was this something which used to work for you, and then stopped?

I never saw this working

## Prerequisites

- Node v16.14.0
- NPM v8.5.1

A node project using npm workspaces running Renovate on GitLab using the GitLab official pipeline template removes from the `package-lock.json` many packages causing `npm ci` to break using node `v16.14.0` and npm `v8.5.1` with the following error

```
npm ERR! `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm ERR! 
npm ERR! Missing: @jcbd/rocky@1.0.0 from lock file
npm ERR! Missing: @jcbd/rocky@1.0.0 from lock file
npm ERR! Missing: @apollo/federation@0.33.8 from lock file
npm ERR! Missing: graphql@15.5.3 from lock file
npm ERR! Missing: @babel/plugin-proposal-class-properties@7.16.7 from lock file
npm ERR! Missing: @babel/plugin-proposal-private-methods@7.16.5 from lock file
npm ERR! Missing: @types/jest@27.0.3 from lock file
npm ERR! Missing: apollo-datasource-rest@3.4.0 from lock file
```

However, the same project is added to GitHub and creates the same MR but the `package-lock.json` contains fewer changes and when the branch is pulled and `npm ci` is executed, it finishes successfully.

### Projects

A reproducible project has been published to GitHub and GitLab to show the bug on GitLab

- GitLab: https://gitlab.com/juancarlosjr97/renovate-rocky-gitlab
- GitHub: https://github.com/juancarlosjr97/renovate-rocky-github

### MR as reference - package `@babel/plugin-proposal-class-properties to v7.16.7`

- GitLab: https://gitlab.com/juancarlosjr97/renovate-rocky-gitlab/-/merge_requests/2
- GitHub: https://github.com/juancarlosjr97/renovate-rocky-github/pull/2

Please see the changes to the `package-lock.json` as it will be clear that there is an issue on GitLab.

### Configuration

Both projects are using the same configuration (if I am correct), both are using the default values as they are not extending the base configuration. The only difference is that the GitLab projects use the [config.js](https://gitlab.com/juancarlosjr97/renovate-rocky-gitlab/-/blob/main/config.js) for Renovate.

#### Relevant debug logs

- GitLab: https://gitlab.com/juancarlosjr97/renovate-rocky-gitlab/-/jobs/2294203324
- GitHub: https://gist.github.com/juancarlosjr97/af5dc79c16077bcc1c0f86462bf42dfc

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
Exactly how are you running on GitLab? e.g. are you using https://gitlab.com/renovate-bot/renovate-runner ?

**Comment by juancarlosjr97** [comment]
If you visit the project you can see I am using the pipeline template https://gitlab.com/juancarlosjr97/renovate-rocky-gitlab/-/blob/main/.gitlab-ci.yml#L6 but I also have tested it using Docker image as follow

```Dockerfile
# Renovate v32.10.3
FROM renovate/renovate@sha256:987823116f895bbcc4c58490f86a8cb5d7b361e84b6ecce04ec78fb30cbcc114

WORKDIR /usr/src/app

COPY config.js config.js

ENV GITHUB_COM_TOKEN=${GITHUB_COM_TOKEN} \
    RENOVATE_TOKEN=${RENOVATE_TOKEN}
```

Both produce the same error on GitLab.

**Comment by viceice** [comment]
Use the dind version, so renovate can respect contraints and will use the right node / npm version for your repo.

**Reply by juancarlosjr97** [reply]
Docker version, that was what I did, right? If so, I already mentioned it here https://github.com/renovatebot/renovate/discussions/14985#discussioncomment-2514018

Otherwise, can you provide an example please @viceice?

Just to clarify, I did not mention it above, but the issue is presently using pipeline templates and docker images on GitLab.

**Reply by viceice** [reply]
we have two templates and you should use the docker-in-docker version, otherwise you'll use the renovate full image, wich will use whatever npm / node version which is installed on that image. So the npm version doesn't match your expected version, so the lockfile get's more changes than expected.

**Reply by viceice** [reply]
https://gitlab.com/renovate-bot/renovate-runner

**Reply by juancarlosjr97** [reply]
Sorry viceice, I am not sure to follow you.

By only doing 

```
include:
    - project: 'renovate-bot/renovate-runner'
      file: '/templates/renovate-dind.gitlab-ci.yml'

services:
    - docker:19.03.15-dind

```

It will respect my node and npm version?

Also, my preference was to use docker image, how could I do it using a docker image?

**Reply by viceice** [reply]
the dind version uses the slim renovate image to run renpovate in docker and then renovate starts the npm lockfile update in a new docker container, where renovate enforces the right node and npm versions to be available

https://docs.renovatebot.com/configuration-options/#constraints

**Reply by juancarlosjr97** [reply]
Ok. I will check that and get back to you.

Thank you @viceice

**Reply by juancarlosjr97** [reply]
@viceice the dind version works great. Now I only have one issue that is not updating the package-lock.json in the monorepo.

I have tried adding `skilInstalls` but still is not update the `package-lock.json` following the documentation https://docs.renovatebot.com/self-hosted-configuration/#skipinstalls. Any suggestion???

**Reply by viceice** [reply]
Please check the debug logs when the commit was made

**Reply by juancarlosjr97** [reply]
Yes, it is the docker in docker error... 

```
DEBUG: Created Merge Request #156 (repository=juancarlosjr97/renovate-rocky-gitlab, branch=renovate/apollo-server-core-3.x)
 WARN: artifactErrors (repository=juancarlosjr97/renovate-rocky-gitlab, branch=renovate/apollo-server-core-3.x)
       "artifactErrors": [
         {
           "lockFile": "package-lock.json",
           "stderr": "docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.\nSee 'docker run --help'.\n"
         }
       ]
```

This is my `.gitlab-ci.yml`

```
---
# yamllint disable-line rule:line-length
image: renovate/renovate:32.6.0-slim@sha256:6a722989e5402841f18b1a488dc3b117fd6afae0f61544134069fd3dba5bba83

variables:
  RENOVATE_BASE_DIR: $CI_PROJECT_DIR/renovate
  RENOVATE_PLATFORM: gitlab
  RENOVATE_ENDPOINT: $CI_API_V4_URL
  GITHUB_COM_TOKEN: $GITHUB_COM_TOKEN
  RENOVATE_TOKEN: $RENOVATE_TOKEN
  LOG_LEVEL: debug

renovate:on-schedule:
  only:
    - schedules
  script:
    - renovate $RENOVATE_EXTRA_FLAGS
```

I am missing something or something else is wrong, but I am not sure what is it. I have all the logs here https://gitlab.com/juancarlosjr97/renovate-rocky-gitlab/-/jobs/2304453734

**Reply by viceice** [reply]
Please follow our templates. We can't support every custom pipeline.

- https://gitlab.com/renovate-bot/renovate-runner
- https://github.com/renovatebot/docker-renovate/blob/main/docs/gitlab.md

---

You are running a docker container to start a renovate container. Why you like to do that?
If you really need to do this, you need to mount a shared ronovate base dir and the docker socket to the renovate container.
So i really suggest to use our pipeline templates / samples as starter.

**Reply by juancarlosjr97** [reply]
Hi @viceice just to clarify the reason for this error. It was because our docker-in-docker was running correctly, but after that it started working as expected :) 

So that was the solution, as we have gitlab self-hosted runners!
